plugins/ipam/ipam.go

```diff
@@ -1061,3 +1068,98 @@ func customIfID(ifName, network string) string {
 	}
 	return ifName + "/" + network
 }
+
+////////////////////////////////////////////////////////////////////////////////
+//    Added By Adel For vni allocation of the SFC vxlan management            //
+////////////////////////////////////////////////////////////////////////////////
+
+// findFirstAvailableID returns the smallest integer that is not present in the
+// given slice of already allocated node IDs.
+func findFirstAvailableVNI(vnis []int) int {
+	res := 5000 // SFC vxlan vni begin from 5000, (low vni <5000 are for other vxlan purpose (like vni=10 for the full-mesched contiv-vpp vxlans)
+	for _, v := range vnis {
+		if res == v {
+			res++
+		} else {
+			break
+		}
+	}
+	return res
+}
+
+// listAllocatedVNIs returns a slice of already allocated SFC VXLAN VNIs (vni>5000).
+func (i *IPAM) listAllocatedVNIs(broker keyval.ProtoBroker) (vnis []int, err error) {
+	vniallocation := &vnialloc.CustomVniVxlanSFC{}
+	it, err := broker.ListKeys(vnialloc.KeyPrefix())
+	if err != nil {
+		return vnis, err
+	}
+
+	for {
+		key, _, stop := it.GetNext()
+		if stop {
+			break
+		}
+
+		// try to read existing vni allocation for the key  (SFC instance)
+
+		found, _, err := broker.GetValue(key, vniallocation)
+		if err != nil {
+			i.Log.Errorf("Unable to read vni allocation: %v", err)
+			//return vnis, err
+		}
+		if !found {
+			i.Log.Errorf("Unable to found vni allocation for key: %v", key)
+		} else {
+			vnis = append(vnis, int(vniallocation.GetVniSfc()))
+		}
+	}
+	return vnis, err
+}
+
+// AllocateVNI tries to allocate a VNI for the SFC instance if not always allocated then persiste it
+func (i *IPAM) AllocateVNI(sfcName string, sfcInsNbr uint32) (vni uint32, err error) {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	key := vnialloc.Key(sfcName, sfcInsNbr)
+	allocation := &vnialloc.CustomVniVxlanSFC{}
+
+	db, err := i.getDBBroker()
+	if err != nil {
+		return 0, err
+	}
+
+	// try to read existing allocation, otherwise create new
```

Comment:
It is not a good practice to look into the database for each VNI allocation request. You should instead store the allocations in memory (e.g. in some map) and only access database in two cases:
- to persist a new allocation (see for example https://github.com/contiv/vpp/blob/dev/plugins/ipam/ipam.go#L617)
- in resync phase to populate the in-memory map with persisted data (for example see  https://github.com/contiv/vpp/blob/dev/plugins/ipam/ipam.go#L254)

Reply:
It is ok. I have created a map (assignedVxlanVNIs map[string]vxlanVniAllocation) that will contain the VNI allocations in memory and resync it from persisted data.